packages/dds/task-manager/src/taskManager.ts

```diff
@@ -758,6 +742,27 @@ export class TaskManagerClass
 		}
 	}
 
+	/**
+	 * Checks whether this client is currently assigned or in queue to become assigned, while also accounting
+	 * for the latest pending ops.
+	 */
+	private queuedOptimistically(taskId: string): boolean {
+		if (this.isAttached() && !this.connected) {
+			return false;
+		}
+
+		assert(this.clientId !== undefined, 0x07f /* "clientId undefined" */);
+
+		const inQueue = this.taskQueues.get(taskId)?.includes(this.clientId) ?? false;
+		const latestPendingOp = this.latestPendingOps.get(taskId);
+		const isPendingVolunteer = latestPendingOp?.type === "volunteer";
+		const isPendingAbandonOrComplete =
+			latestPendingOp?.type === "abandon" || latestPendingOp?.type === "complete";
+		// We return true if the client is either in queue already or the latest pending op for this task is a volunteer op.
+		// But we should always return false if the latest pending op is an abandon or complete op.
+		return (inQueue || isPendingVolunteer) && !isPendingAbandonOrComplete;
```

Comment:
Nit - this grouping is a little odd since isPendingVolunteer is mutually exclusive with isPendingAbandonOrComplete (so the check is redundant in some cases), might read a little more naturally as
```suggestion
		return (inQueue && !isPendingAbandonOrComplete) || isPendingVolunteer;
```